### PR TITLE
Scenario template editor

### PR DESCRIPTION
**Pull request by NickAragua**
Take 2 of the Scenario Template Editor feature.

It's quite extensive, and, at the current moment, can only be invoked "manually". The next phase of the AtB update will hook it up to the AtB system.

Main features:
Scenario Template Editor
templateeditor

This allows the user to define a scenario template, which is a collection of forces, various generation and arrival mechanisms and map parameters.

"Generate From Template"
In the briefing tab, "add scenario" then "generate from template" to create a scenario from a defined scenario template. You can also apply modifiers from a fixed list for greatly increased variety.

Field guns and artillery tubes now show up in personnel market / unit market respectively. Infantry opfor may contain field guns as well.

### CODE REVIEW

File: MekHQ/src/mekhq/campaign/mission/ScenarioForceTemplate.java [hunk 1]
```diff
@@ -0,0 +1,517 @@
...
+    
+    public enum ForceAlignment {
+        Player,
+        Allied,
+        Opposing,
+        Third,
+        PlanetOwner;
+        
+        public static ForceAlignment getForceAlignment(int ordinal) {
+            for (ForceAlignment he : values()) {
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/campaign/mission/ScenarioForceTemplate.java]: `he`?

File: MekHQ/src/mekhq/campaign/mission/ScenarioForceTemplate.java [hunk 2]
```diff
@@ -0,0 +1,517 @@
...
+    public ScenarioForceTemplate() {
+        
+    }
+    
+    public ScenarioForceTemplate(int forceAlignment, int generationMethod, double forceMultiplier, List<Integer> deploymentZones,
+            int destinationZone, int retreatThreshold, int allowedUnitType) {
+        this.forceAlignment = forceAlignment;
+        this.generationMethod = generationMethod;
+        this.forceMultiplier = forceMultiplier;
+        this.deploymentZones = deploymentZones == null ? new ArrayList<>() : deploymentZones;
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/campaign/mission/ScenarioForceTemplate.java]: preference: copy `deploymentZones` to local ArrayList<>

File: MekHQ/src/mekhq/campaign/mission/ScenarioMapParameters.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        AllGroundTerrain,
+        SpecificGroundTerrain,
+        Space,
+        LowAtmosphere
+    }
+    
+    /**
+     * The fixed base width/x dimension of the map.
+     */
+    public int baseWidth;
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/campaign/mission/ScenarioMapParameters.java]: prefer getters/setters, JAXB will handle

File: MekHQ/src/mekhq/campaign/mission/ScenarioObjective.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+ * @author NickAragua
+ *
+ */
+public class ScenarioObjective {
+    public enum ObjectivePriority {
+        Primary,
+        Secondary
+    };
+    
+    public ObjectivePriority priority;
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/campaign/mission/ScenarioObjective.java]: getters/setters

File: MekHQ/src/mekhq/campaign/mission/ScenarioTemplate.java [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+ * This is the root data structure for organizing information related to a scenario template.
+ * @author NickAragua
+ *
+ */
+@XmlRootElement(name="ScenarioTemplate")
+public class ScenarioTemplate {
+    public static final String ROOT_XML_ELEMENT_NAME = "ScenarioTemplate";
+    
+    
+    public String name;
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/campaign/mission/ScenarioTemplate.java]: getters/setters

File: MekHQ/src/mekhq/campaign/mission/ScenarioTemplate.java [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+     * @param inputFile The source file
+     * @return Possibly an instance of a ScenarioTemplate
+     */
+    public static ScenarioTemplate Deserialize(File inputFile) {
+        ScenarioTemplate resultingTemplate = null;
+        
+        try {
+            JAXBContext context = JAXBContext.newInstance(ScenarioTemplate.class);
+            Unmarshaller um = context.createUnmarshaller();
+            JAXBElement<ScenarioTemplate> templateElement = um.unmarshal(new StreamSource(inputFile), ScenarioTemplate.class);
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/campaign/mission/ScenarioTemplate.java]: don't use StreamSource, use `MekHqXmlUtil.createSafeXmlSource`

File: MekHQ/src/mekhq/campaign/mission/atb/AtBScenarioModifier.java [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+     * Possible values for when a scenario modifier may occur: before or after primary force generation.
+     * @author NickAragua
+     *
+     */
+    public enum EventTiming {
+        PreForceGeneration,
+        PostForceGeneration
+    }
+
+    public String modifierName = null;
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/campaign/mission/atb/AtBScenarioModifier.java]: getters/setters

File: MekHQ/src/mekhq/campaign/mission/atb/AtBScenarioModifier.java [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+        try {
+            JAXBContext context = JAXBContext.newInstance(AtBScenarioModifier.class);
+            Unmarshaller um = context.createUnmarshaller();
+            File xmlFile = new File(fileName);
+            if(!xmlFile.exists()) {
+                MekHQ.getLogger().warning(AtBScenarioModifier.class, "Deserialize", String.format("Specified file %d does not exist", fileName));
+                return null;
+            }
+            
+            JAXBElement<AtBScenarioModifier> templateElement = um.unmarshal(new StreamSource(xmlFile), AtBScenarioModifier.class);
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/campaign/mission/atb/AtBScenarioModifier.java]: use createSafeXmlSource here too

File: MekHQ/src/mekhq/campaign/mission/atb/AtBScenarioModifier.java [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+        try {
+            JAXBContext context = JAXBContext.newInstance(ScenarioModifierManifest.class);
+            Unmarshaller um = context.createUnmarshaller();
+            File xmlFile = new File(fileName);
+            if(!xmlFile.exists()) {
+                MekHQ.getLogger().warning(ScenarioModifierManifest.class, "Deserialize", String.format("Specified file %s does not exist", fileName));
+                return null;
+            }
+            
+            JAXBElement<ScenarioModifierManifest> templateElement = um.unmarshal(new StreamSource(xmlFile), ScenarioModifierManifest.class);
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/campaign/mission/atb/AtBScenarioModifier.java]: createSafeXmlSource

File: MekHQ/src/mekhq/campaign/universe/IUnitGenerator.java [hunk 10]
```diff
@@ -44,6 +45,23 @@
             EntityMovementMode.WHEELED, EntityMovementMode.HOVER, EntityMovementMode.WIGE,
             EntityMovementMode.VTOL);
     
+    /**
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/campaign/universe/IUnitGenerator.java]: whitespace off

File: MekHQ/src/mekhq/campaign/mission/AtBDynamicScenarioFactory.java [hunk 11]
```diff
@@ -0,0 +1,1741 @@
...
+        String crewName = rng.generate();
+
+        RandomSkillsGenerator rsg = new RandomSkillsGenerator();
+        rsg.setMethod(RandomSkillsGenerator.M_TAHARQA);
+        rsg.setLevel(skill);
+
+        if (f.isClan()) {
+            rsg.setType(RandomSkillsGenerator.T_CLAN);
+        }
+        int[] skills = rsg.getRandomSkills(en);
```
[the remark this answers is not in the document]
Comment (NickAragua) [reply MekHQ/src/mekhq/campaign/mission/AtBDynamicScenarioFactory.java]: True, though this is a pre-existing method, so I'd probably like to leave it as is for the time being as it's used in many other places.

File: MekHQ/src/mekhq/campaign/mission/AtBDynamicScenarioFactory.java [hunk 12]
```diff
@@ -0,0 +1,1741 @@
...
+        if (normalizedWeight < 80) {
+            return EntityWeightClass.WEIGHT_HEAVY;
+        }
+        if (normalizedWeight < 100) {
+            return EntityWeightClass.WEIGHT_ASSAULT;
+        }
+        return EntityWeightClass.WEIGHT_SUPER_HEAVY;
+    }
+    
+    /*private static void generateBasicForceUnit(String faction, int unitType, int weightClass, int year, int skill, int rating) {
```
[the remark this answers is not in the document]
Comment (NickAragua) [reply MekHQ/src/mekhq/campaign/mission/AtBDynamicScenarioFactory.java]: I was experimenting with using the force generator here. I kind of want to keep it around to retain what I learned from that.

File: MekHQ/src/mekhq/campaign/mission/AtBDynamicScenarioFactory.java [hunk 13]
```diff
@@ -0,0 +1,1741 @@
...
+        } catch (Exception ex) {
+            MekHQ.getLogger().log(AtBDynamicScenarioFactory.class, METHOD_NAME, LogLevel.ERROR,
+                    "Unable to load unit: " + name); //$NON-NLS-1$
+            MekHQ.getLogger().error(AtBDynamicScenarioFactory.class, METHOD_NAME, ex);
+        }
+        if (mechFileParser == null) {
+            return null;
+        }
+
+        Entity en = mechFileParser.getEntity();
```
[the remark this answers is not in the document]
Comment (NickAragua) [reply MekHQ/src/mekhq/campaign/mission/AtBDynamicScenarioFactory.java]: Before where?

File: MekHQ/src/mekhq/campaign/mission/AtBDynamicScenarioFactory.java [hunk 14]
```diff
@@ -0,0 +1,1741 @@
...
+     * @param campaign Working campaign
+     * @return List of generated entities.
+     */
+    private static List<Entity> generateLance(String faction, int skill, int quality, List<Integer> unitTypes, 
+            String weights, boolean artillery, Campaign campaign) {
+        List<Entity> retval = new ArrayList<>();
+        
+        for(int i = 0; i < unitTypes.size(); i++) {
+            Entity en = getEntity(faction, skill, quality, unitTypes.get(i), 
+                    AtBConfiguration.decodeWeightStr(weights, i), artillery, campaign);
```
[the remark this answers is not in the document]
Comment (NickAragua) [reply MekHQ/src/mekhq/campaign/mission/AtBDynamicScenarioFactory.java]: They should, but I added some safety checking.

File: MekHQ/src/mekhq/campaign/mission/AtBDynamicScenarioFactory.java [hunk 15]
```diff
@@ -0,0 +1,1741 @@
...
+            setWeather(scenario);
+        }
+        
+        if(campaign.getCampaignOptions().getUsePlanetaryConditions()) {
+            setPlanetaryConditions(scenario, contract, campaign);
+        }
+        
+        setTerrain(scenario);
+        
+        //setScenarioModifiers(scenario);
```
[the remark this answers is not in the document]
Comment (NickAragua) [reply MekHQ/src/mekhq/campaign/mission/AtBDynamicScenarioFactory.java]: Initially I wanted to apply random scenario modifiers to an instance of a scenario template, but that feature needs a little more work.

**Pull request merged**